src/main/java/com/vinylteam/vinyl/dao/jdbc/JdbcRecoveryPasswordDao.java

```diff
@@ -0,0 +1,114 @@
+package com.vinylteam.vinyl.dao.jdbc;
+
+import com.vinylteam.vinyl.dao.RecoveryPasswordDao;
+import com.vinylteam.vinyl.dao.jdbc.mapper.RecoveryRowMapper;
+import com.vinylteam.vinyl.entity.RecoveryToken;
+import com.zaxxer.hikari.HikariDataSource;
+import lombok.extern.slf4j.Slf4j;
+import org.postgresql.util.PSQLException;
+
+import java.sql.*;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.util.Optional;
+
+@Slf4j
+public class JdbcRecoveryPasswordDao implements RecoveryPasswordDao {
+
+    private static final String INSERT_TOKEN = "INSERT INTO recovery_password" +
+            " (user_id, token, created_at)" +
+            " VALUES (?, ?, ?)" +
+            " ON CONFLICT (user_id) DO UPDATE SET token = ?, created_at = ?";
+    private static final String FIND_BY_TOKEN = "SELECT id, user_id, token, created_at FROM recovery_password" +
+            " WHERE token = ?";
+    private static final String REMOVE_TOKEN = "DELETE FROM recovery_password" +
+            " WHERE token = ?";
+
+    private final RecoveryRowMapper recoveryRowMapper = new RecoveryRowMapper();
+    private final HikariDataSource dataSource;
+
+    public JdbcRecoveryPasswordDao(HikariDataSource dataSource) {
+        this.dataSource = dataSource;
+    }
+
+    @Override
+    public boolean addRecoveryUserToken(RecoveryToken recoveryToken) {
+        boolean isAdded = false;
+        try (Connection connection = dataSource.getConnection();
+             PreparedStatement addTokenStatement = connection.prepareStatement(INSERT_TOKEN)) {
+            addTokenStatement.setLong(1, recoveryToken.getUserId());
+            addTokenStatement.setString(2, recoveryToken.getToken());
+            addTokenStatement.setTimestamp(3, Timestamp.from(Instant.now()));
+            addTokenStatement.setString(4, recoveryToken.getToken());
+            addTokenStatement.setTimestamp(5, Timestamp.from(Instant.now()));
+            log.debug("Prepared statement {'preparedStatement':{}}.", addTokenStatement);
+            int result = addTokenStatement.executeUpdate();
+            if (result > 0) {
+                isAdded = true;
+            }
+        } catch (PSQLException e) {
+            log.error("Database error while adding recovery token to recovery_password", e);
+            isAdded = false;
```

Comment:
I tried to get rid of this, but without success, PSQLException comes from postgres, sql - from java, now only SQLException will be catched

Reply:
> When we were adding user to DB, we have got PSQLException if user with the same login existed and SqlException in case of other problems. It is rude decision but it gave us approtunite to divide exceptions into two cases and create logic depending on this. In this case we can catch only SQL exceptions.

Reply:
PSQL - database limitations: not duplicate uuid token and user id.
It's good for user dao where user tried to sign up with same email or something - to return false when not added.

But there aren't any acceptable situations like that with tokens - so if we can't add token exception should be thrown.

Reply:
testcase for "other problem" doesn't exists, so I can change to SQLException and all tests work

Reply:
Nastya, then they shouldn't return boolean, I tried to not change DAO, and services, test already prepared to work with this, so I wouldn't change your logic, just for user is important - was it added or not, if not - what is difference about exception?